Health Check for Redis

ingvaldlorentzen opened the pull request:
This adds support for a health check for Redis. 
More or less the same as the health check for RabbitMQ. 

It currently uses the `celery_broker_url` from Django settings. This could be changed to `broker_url` that RabbitMQ uses, if wanted.

ingvaldlorentzen committed:
- Added Redis health check, inspired by the RabbitMQ healthcheck
- Fixed typo in RedisMQ logger message

ingvaldlorentzen committed:
Added missing `default_app_config` in `__init__.py`

ingvaldlorentzen committed:
Fixer wrong attribute value in redis test

codingjoe commented:
@ingvaldlorentzen thanks for the contribution. I think a Redis health-check would be great. That being said I see multiple places where Redis can be used in Django. I for one, use it for caching, not as a result backend for celery. So either we change implementation, or we change the naming a little to reflect, that it's monitoring the result backend.

Please also drop the commented out lines ;)

Thanks!
Joe

ingvaldlorentzen committed:
- Removed commented code
- Generalized Django settings Redis URL name
- Updated variable naming for redis health check

ingvaldlorentzen commented:
Hi, 

Thanks for the feedback!
I changed the Redis URL from `CELERY_BROKER_URL` to simply `REDIS_URL`. Also updated the variables in the health check to reflect this change. 
I think `REDIS_URL ` is a nice generic way to do this, and if a project doesn't use `REDIS_URL`, it's easy and hopefully understandable to add to the settings variables for the health check. 
Removed the commented code, sorry about that, it was a slip-up.

I'll add info on the Redis health check to the documentation if you think this looks alright. 

Regards, 
Ingvald

**Edit**: I see some checks are failing here. I'm not super familiar with codecov, if you have any feedback on how I should proceed to fix this I would be super thankful!
If not, I'll see if I can get a team member to help me out on Monday.

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,37 @@
+import logging
+
+from django.conf import settings
+from kombu import Connection
```
`kombu` is part of celery and current not listed in the requirements. So maybe it makes sense to switch to another redis client. There also should be documentation, or an `extra` requirement addtion to the setuptools config.

ingvaldlorentzen replied:
Added some documentation to the README, not a whole lot, but not sure if any more is really necessary.

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,37 @@
+        except ConnectionRefusedError as e:
+            self.add_error(ServiceUnavailable("Unable to connect to Redis: Connection was refused."), e)
+
+        except IOError as e:
```
You should be more verbose about what error you are expecting here.

JonasKs replied:
I removed it entirely, I think the main reason for it to be there was due to the `RabbitMQ` backend had it. Not sure why it would be Excepted specifically. Please let me know if you want it added back.

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,37 @@
+        except IOError as e:
+            self.add_error(ServiceUnavailable("IOError"), e)
+
+        except BaseException as e:
```
Please don't catch base exceptions, those are already handled by the `BaseHealthCheckBackend`

JonasKs replied:
Fixed.

codingjoe commented on tests/test_redis.py:
```diff
@@ -0,0 +1,30 @@
+from health_check.contrib.redis.backends import RedisHealthCheck
+
+
+class TestRabbitMQHealthCheck:
```
```suggestion
class TestRedisHealthCheck:
```
;)

JonasKs replied:
Fixed.

codingjoe commented on tests/test_redis.py:
```diff
@@ -0,0 +1,30 @@
+        conn_exception = ConnectionRefusedError("Refused connection")
+
+        # mock returns
+        mocked_conn = mock.MagicMock()
```
You test only one exception type here. Please test all of them.

JonasKs replied:
Fixed. 
I see `patch` and `project` still fails. I can't see `rabbitmq.apps` being tested, so I'm not sure if you want this to be tested too?

codingjoe reviewed (changes requested):
@ingvaldlorentzen I like the changes you made. There certainly needs to be some documentation added for the setting and health check as a whole. I think this is on a good path :) Keep up the good work!

JonasKs committed:
Added tests for all exceptions. Removed IO Exception catch. Added kombu exceptions handling. Removed BaseException as requested by Codingjoe.

JonasKs committed:
Fixed how imports of exceptions is done.

JonasKs committed:
Had to fix sorting of imports, Travis is picky.

JonasKs commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,36 @@
+        except exceptions.ConnectionLimitExceeded as e:
+            self.add_error(ServiceUnavailable("Unable to connect to Redis: "
+                                              "Maximum number of simultaneous connections exceeded"), e)
+        else:
```
Hard to have codecov on this, since there's nothing to be returned or anything.

codingjoe replied:
Just write a test case, where nothing fails. You can use py.test's [caplog](https://docs.pytest.org/en/latest/logging.html) to make assertions on the debug log.

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,36 @@
+import logging
+
+from django.conf import settings
+from kombu import Connection, exceptions
```
`kombu` is still not listed as a requirement. I also don't think that you need `kombu` since it's a messaging library. I would prefer to use plain [redis](https://pypi.org/project/redis/).

ingvaldlorentzen replied:
Redis library implemented instead of kombu

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,36 @@
+            self.add_error(ServiceUnavailable("Unable to connect to Redis: "
+                                              "Maximum number of simultaneous connections exceeded"), e)
+        else:
+            logger.debug("Connection established. Redis is healthy.")
```
I would use `info` over `debug` here.

codingjoe reviewed (changes requested):
Almost there! Keep up the good work @ingvaldlorentzen

ingvaldlorentzen committed:
- Changed redis connection to use redis package instead of kombu
- Added exceptions to reflect connection change
- Bumped logging to info

ingvaldlorentzen committed:
Added basic info on redis healthcheck to README

JonasKs committed:
Added tests for new redis usage. Tox QA is behaving strangely, going to let Travis run it once.

JonasKs committed:
A newline slipped..

JonasKs committed:
Removed prints, oops

JonasKs commented:
I tried to create the tests for most things now that @ingvaldlorentzen has refactored over to use `redis`.

I haven't used pytest before, so please give me a heads up if I need to correct something.  
`test_redis_conn_ok` could probably be rewritten to catch the log message as you requested, but I'd have to look more into that in two weeks when I'm back from vacation.

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,37 @@
+        """Check Redis service by opening and closing a connection."""
+        logger.debug("Checking for a redis_url in django settings...")
+
+        redis_url = getattr(settings, "REDIS_URL", None)
```
Can you make that a class attribute? I for one, have multiple Redis servers running, so I will need to subclass this health check.

ingvaldlorentzen replied:
Good idea! Added it

codingjoe commented on health_check/contrib/redis/backends.py:
```diff
@@ -0,0 +1,37 @@
+        except BaseException as e:
+            self.add_error(ServiceUnavailable("Unknown error"), e)
+        else:
+            logger.info("Connection established. Redis is healthy.")
```
```suggestion
            logger.debug("Connection established. Redis is healthy.")
```

ingvaldlorentzen replied:
Fixed.

codingjoe reviewed (changes requested):
Looks very good,  I just had some final comments.

ingvaldlorentzen committed:
- Fixed logging level
- Added `redis_url` as class attribute for subclassing

ingvaldlorentzen committed:
- Moved class attributes to `__init__` for better mocking in tests

ingvaldlorentzen committed:
- Alphabetized imports

ingvaldlorentzen committed:
- Added period to docstring

codingjoe committed:
Minor changes

codingjoe commented:
Released in [3.11.0](https://github.com/KristianOellegaard/django-health-check/releases/tag/3.11.0) 🎉
